golang/cosmos/x/swingset/keeper/store_snapshotter.go

```diff
@@ -0,0 +1,685 @@
+//   the snapshot operation has started.
+// - Some components, in particular state-sync, may need to perform other work
+//   alongside taking a swing-store snapshot. This work similarly cannot block
+//   the main execution, but must allow for the swing-store synchronization
+//   that enables deterministic snapshots. For state-sync, this work happens
+//   before the swing-store snapshot can be consumed.
+//
+// The general approach taken is to implement a SwingStoreSnapshotter that
+// implements the communication with the JS side, enforces that no concurrent
+// operations take place, defers the consumption of the snapshot to a provided
+// SwingStoreSnapshotTaker, and provides some synchronization methods to let
+// the component performing the snapshot enforce mutation boundaries.
+//
+// To simplify the implementation, we require a single instance of
+// SwingStoreSnapshotter exists (not enforced), and that all method calls be
+// performed from the same thread (no mutex guard).
+//
+// The process of taking a snapshot proceeds as follow:
+// - The component invokes swingStoreSnapshotter.InitiateSnapshot with a taker
+// - InitiateSnapshot verifies no other snapshot operation is in progress and
+//   starts a goroutine to perform the snapshot operation. It requests the JS
+//   side to start taking a snapshot of the swing-store, and calls the taker's
+//   TakeSnapshot method with a function param allowing it to retrieve the
+//   snapshot.
+// - The cosmos app with call swingStoreSnapshotter.WaitUntilSnapshotStarted
+//   before instructing the JS controller to commit its work, satisfying the
+//   deterministic snapshot requirement.
+// - TakeSnapshot must call the retrieve function before returning, however it
+//   may perform other work before. For cosmos state-sync snapshots,
+//   TakeSnapshot will perform its own snapshotting logic before reaching the
+//   "swingset extension"'s snapshotter that will retrieve and process the
+//   swing-store snapshot.
+// - When the retrieve function is called, it blocks until the JS snapshot is
+//   ready, then creates a SwingStoreSnapshotProvider that abstract accessing
+//   the content of the snapshot. The taker's ProcessExportedSnapshot is called
+//   with the provider
+// - ProcessExportedSnapshot reads the snapshot using the provider.
+//
+// Restoring a swing-store snapshot does not have similar non-blocking
+// requirements. The component simply invokes swingStoreSnapshotter.RestoreSnapshot
+// with a SwingStoreSnapshotProvider representing the swing-store snapshot to
+// be restored, and RestoreSnapshot will consume it and block until the JS side
+// has completed the restore before returning.
+
+// ExportManifestFilename is the manifest filename which must be synchronized with the JS export/import tooling
+// See packages/cosmic-swingset/src/export-kernel-db.js and packages/cosmic-swingset/src/import-kernel-db.js
+const ExportManifestFilename = "export-manifest.json"
```

Comment:
There should be a similar cross-reference pointing here from packages/cosmic-swingset/src/export-kernel-db.js.